### ISSUE TITLE
Generating lastname from TO name

### ISSUE THREAD

**Issue by jrathlev** [thundernest/quicktext#96]
If a TO name is given in the form "Meier, Max" and the _lastname_ is inserted using Quicktext, _lastname_ and _firstname_ are swapped:
_lastname_ = Max
_firstname_ = Meier,

Checking the source code in _wzQuicktextVar.jsm_ (_function process_to_), I found out that Quicktext begins to compute the names in a correct way (depending, if there is a comma as separator). Then it calls the function _getCardForEmail_ to check if there is a card entry. But this card entry was generated by Thunderbird in a wrong way (lastname and firstname are swapped). Quicktext replaces the correct values computed before with wrong values from the card entry.

**Comment by jobisoft** [comment]
That is a tough one. If I change that, the other half of the planet will complain (those where the strings are swapped in the TO header and correct in the TB addressbook).

Can't you just sanitize your address book?

**Comment by jrathlev** [comment]
Unfortunately it is not reasonable to keep the address book up to date. This would require manual editing of each new mail address. I am receiving many mail queries from different senders and Thunderbird automatically generates a new card entry for each. I think that it is the task of the Thunderbird developers to improve this function.

But for now, it would be helpful to have an option in the settings of Quicktext to exclude looking up the card values, preferably separate for each mail account.

**Comment by jobisoft** [comment]
Hm. Thunderbird is not adding addresses of incoming emails to any addressbook. It adds addresses of outgoing emails to the "collected addresses" address book (and that can be disabled) and in my case it does that correct. So something is wrong, but not with quicktext.

What addressbook is quicktext looking for cards? Can you point me to the source?

Do you have any addon ionstalled, that could cause the behaviour (adding addresses of incoming email)?

**Comment by jobisoft** [comment]
Ah, now I got it. If I send an email to

`Maier, Hans <mh@something.org>`

Thunderbird will add it wrong indeed, "Maier," as first name, "Hans" as last name. I never did that before. All my contacts use the "normal" order.

**Comment by jrathlev** [comment]
Replying to a mail from a new sender for the first time, Quicktext generates the correct lastname. But on sending, Thunderbird adds the address to the "collected  addresses" and generates the wrong card entry. 
When replying a second time to the same sender, Quicktext uses this entry for lastname. 
Disabling to add outgoing emails to the address book is unfortunately a global setting and cannot be restricted to selected mail accounts.
In Germany it is common to use the reverse order with a comma as separator.

**Comment by jobisoft** [comment]
How about this: I do not add a UI option as that would be a lot of work. Lets add a check on the data returned from the card and if the data looks swapped ("," at the end of the first name) we swap it back before passing it on.

Can you contribute?

**Comment by jrathlev** [comment]
This is a good and simple proposal

**Comment by jobisoft** [comment]
Can you test this version? Create a new profile and install it via the gear menu in the Thunderbird  AddOn Manager.

If you do not want to create a new profile, you need to uninstall your current quicktext, restart twice (!) and than install the attached version. TB68 has a addon caching issue.

[quicktext.xpi](https://github.com/thundernest/quicktext/files/3907698/quicktext.zip)

**Comment by jrathlev** [comment]
I tested the new version and it works correctly. Am I right assuming that the values from the card entry now only will be used. if there was no name found in th TO address?

**Comment by jobisoft** [comment]
No, I kept the original logic, that the value from the card overides the value in the TO header. Usually these two are the same (modulo swapping) but the user may change it and thus the user defined value should be used.

What I did is to move the swapping logic already used for the header to the very end of the function and let it act on the final fullname (either from header or from card if present)

What I also kept is, that only the firstname + lastname value from the card is used, not the displayname (which remains swapped if you add that to your template)

**Comment by jrathlev** [comment]
No, I do not think so. At first the fullname is taken from the header. Only if this is empty, the fullname will be replaced by firstname + lastname from the card.
Another small issue: firstname should be trimmed to remove leading space.

**Comment by jobisoft** [comment]
> At first the fullname is taken from the header. Only if this is empty, the fullname will be replaced by firstname + lastname from the card.

That is not how itis in the code:

https://github.com/thundernest/quicktext/blob/a070316a1b2d01b97448168153525ffa7d68a66d/chrome/content/modules/wzQuicktextVar.jsm#L692-L710

Could you contribute the nit picking yourself?

**Comment by jobisoft** [comment]
Well, looking at the code again, I need to confirm that the flow indeed changed. As fullname is not a true property and thus is not overwritten by the card property iterator, it will keep the original header value. Good catch.

**Comment by jobisoft** [comment]
I think I removed the flaw and also did the trimming, while i was at it. New version attached.

[quicktext.xpi](https://github.com/thundernest/quicktext/files/3908254/quicktext.zip)

**Comment by jrathlev** [comment]
I tested the new version and everything seems to work as it should.

**Comment by jobisoft** [comment]
Perfect. I will release it to ATN after #75 has landed:

FYI: You should not go back to the current official version, as the version linked above changed the type of a pref from int to char and the "current" version does not have a check for it and will probably error out if it wants to store its int value in the now char typed pref value.

**Comment by jrathlev** [comment]
There is a small flaw left.  If the card query returns an empty fullname, the following name swapping fails. So there should be an additional if statement:
`var name = this.mData['TO'].data['fullname'][k];`
`if (!name) this.mData['TO'].data['fullname'][k] = TrimString(names.value[i]); `
`if (name)  ...`

**Comment by jobisoft** [comment]
[quicktext.xpi](https://github.com/thundernest/quicktext/files/3919191/quicktext.zip)

Can you test the above version?

**Comment by jobisoft** [comment]
Deleted last post, little modification. Can you test?

[quicktext.xpi](https://github.com/thundernest/quicktext/files/3919203/quicktext.zip)

**Comment by jrathlev** [comment]
I tested the new version but now there is no text (also static text) inserted anymore.

**Comment by jobisoft** [comment]
[quicktext_2.7.1.xpi](https://github.com/thundernest/quicktext/files/3921361/quicktext.zip)

Can you retry with this version?

**Comment by jrathlev** [comment]
Now everything is ok. I tested several different mail headers.
